### PR TITLE
Add hosted MCP tools with `GoogleWorkspace`

### PR DESCRIPTION
**Pull request by adtyavrdhn**
Connect an agent to the provider's MCP tools with `GoogleWorkspace`:

```python
from pydantic_ai import Agent
from pydantic_ai_harness.google_workspace import GoogleWorkspace

agent = Agent('openai:gpt-5.6-sol', capabilities=[GoogleWorkspace(services=['calendar'])])
```

Select Google Workspace services and authenticate with a caller-provided OAuth access token or HTTP auth. Service prefixes keep tool names distinct.

The common options are `auth`, `read_only`, and `include_instructions`. Provider credentials and server settings determine access; the default exposes the server's available tools, including writes. Applications compose existing toolset filtering and approval wrappers when needed.

Both documentation pages describe setup, read-only behavior, approval, and connection ownership. Focused tests exercise tool execution and instruction forwarding through `Agent(TestModel(...))` with an in-process MCP server, plus connection configuration. These are not authenticated provider smoke tests.

Validation: focused capability tests, combination tests, docs parity, Ruff, and file-scoped Pyright pass. Full CI runs on the PR.

Relates to #787.

### CODE REVIEW

**Comment by adtyavrdhn** [comment]
Acknowledged. The sensitive-data boundary is why reads are the default, mutations require explicit opt-in, credentials are excluded from specs and representations, and hosted applications can supply caller-owned clients. Human review remains appropriate.

**Comment by adtyavrdhn** [comment]
@codex review

**Comment by adtyavrdhn** [comment]
@codex review

**Comment by adtyavrdhn** [comment]
@codex review

File: pydantic_ai_harness/__init__.py [hunk 1]
```diff
@@ -76,6 +77,7 @@
     'FileSystem',
     'GuardrailError',
     'GuardrailResult',
+    'GoogleWorkspace',
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/__init__.py]: Fixed in 2b76b583. The root export is removed and both examples now import from `pydantic_ai_harness.google_workspace`.

File: docs/google-workspace.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
+# Google Workspace
+
+Google Workspace gives an agent selected Gmail, Calendar, Drive, Docs, Sheets, Slides, Chat, and People tools through Google's official remote MCP servers. Gmail and Calendar are selected by default.
+
+[Source](https://github.com/pydantic/pydantic-ai-harness/tree/main/pydantic_ai_harness/google_workspace/)
+
+> While Pydantic AI Harness is on 0.x releases, the API may change between minor releases; when it does, the release notes explain the migration. See the [version policy](../docs/index.md#version-policy).
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply docs/google-workspace.md]: Fixed in 2b76b583. The site page now uses the canonical stability text and `index.md#version-policy` link.

File: docs/google-workspace.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+import os
+
+from pydantic_ai import Agent
+from pydantic_ai_harness import GoogleWorkspace
+
+workspace = GoogleWorkspace(read_only=False, access_token=os.environ['GOOGLE_ACCESS_TOKEN'])
+agent = Agent(
+    'openai:gpt-5.6-sol',
+    toolsets=[workspace.get_toolset().approval_required()],
+    defer_model_check=True,
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply docs/google-workspace.md]: The optional approval example was removed when the docs were reduced to one complete runnable read-only example, so the invalid snippet no longer exists.

File: docs/google-workspace.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+5. If you select Chat, configure the Google Chat API app and turn off **Enable interactive features**.
+
+| Product | Scope suffixes |
+|---|---|
+| Gmail | `gmail.readonly`, `gmail.compose` |
+| Drive | `drive.readonly`, `drive.file` |
+| Docs | `drive.readonly`, `drive.file`, `documents.readonly`, `documents` |
+| Sheets | `drive.readonly`, `drive.file`, `spreadsheets.readonly`, `spreadsheets` |
+| Slides | `drive.readonly`, `drive.file`, `presentations.readonly`, `presentations` |
+| Calendar | `calendar.calendarlist.readonly`, `calendar.events.freebusy`, `calendar.events.readonly` |
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply docs/google-workspace.md]: Fixed in 2b76b583. Both docs now separate default read-only scopes from additional mutation scopes, including `gmail.modify` and `calendar.events`.

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            if unused_clients:
+                raise UserError(f'Client configured for unselected service: {sorted(unused_clients)[0]!r}.')
+
+        if isinstance(self.allowed_tools, str):
+            self.allowed_tools = (self.allowed_tools,)
+        elif self.allowed_tools is not None:
+            self.allowed_tools = tuple(self.allowed_tools)
+        if self.allowed_tools is not None:
+            selected_prefixes = tuple(f'{service}_' for service in services)
+            invalid = next((name for name in self.allowed_tools if not name.startswith(selected_prefixes)), None)
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: Fixed in 00d97dad. Runtime members are validated before prefix matching, with a public regression test for malformed values.

[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: Follow-up in 37423680 uses a unique sentinel so `None` is also rejected; the regression test covers both integer and null members.

**Comment by adtyavrdhn** [comment]
@codex review

**Comment by adtyavrdhn** [comment]
@codex review

File: docs/google-workspace.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- "Summarize unread project mail and list my meetings today."
+- "Find the launch plan in Drive, read the linked Doc, and show the budget values from its Sheet."
+- "Read the quarterly Slides presentation and list the files I can share with the team."
+- "Search Chat for the release decision and find the email address for its author in People."
+
+## Operational constraints
+
+- Gmail and Calendar are selected by default. Pass `services=('drive', 'docs', 'sheets', 'slides', 'chat', 'people')` to select other products.
+- Tool names are prefixed by service. Use `allowed_tools=('gmail_search_threads', 'calendar_list_events')` for an exact allowlist.
+- `read_only=True` is the default. It filters exposed tools, but it does not reduce the bearer token's scopes. Issue the token with the narrowest scopes your application permits. Set `read_only=False` to expose server-provided mutation tools, then use `workspace.get_toolset().approval_required()` when those actions need human approval.
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply docs/google-workspace.md]: Fixed in b50edcba. Both docs now give the complete public wiring: pass the approval-wrapped toolset through `toolsets` instead of retaining the capability, and include `DeferredToolRequests` in `output_type`. The focused docs parity and runnable snippet selection passes.

**Comment by adtyavrdhn** [comment]
@codex review

File: docs/google-workspace.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply docs/google-workspace.md]: Final follow-up in 2314f807 also preserves `workspace.get_instructions()` in the approval Agent configuration and accurately states that the default wrapper defers every exposed operation. A behavioral assertion verifies that the untrusted-content guidance reaches the model.

**Comment by adtyavrdhn** [comment]
@codex review

**Comment by adtyavrdhn** [comment]
@codex review

**Comment by adtyavrdhn** [comment]
@codex review

File: pydantic_ai_harness/__init__.py [hunk 7]
```diff
@@ -130,6 +132,7 @@
     'ExaSearch': 'exa',
     'FallbackCompaction': 'compaction',
     'FileSystem': 'filesystem',
+    'GoogleWorkspace': 'google_workspace',
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/__init__.py]: Fixed in `a7250e1b`. The root `TYPE_CHECKING` import, `__all__` item, and lazy dispatch entry are removed, and both runnable examples again import from `pydantic_ai_harness.google_workspace`. Focused capability/combine/docs-parity tests pass (405), both snippets pass, explicit-path Pyright reports 0 errors, and Ruff is clean.

**Comment by adtyavrdhn** [comment]
@codex review

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    'sheets': 'https://sheetsmcp.googleapis.com/mcp/v1',
+    'slides': 'https://slidesmcp.googleapis.com/mcp/v1',
+    'calendar': 'https://calendarmcp.googleapis.com/mcp/v1',
+    'chat': 'https://chatmcp.googleapis.com/mcp/v1',
+    'people': 'https://people.googleapis.com/mcp/v1',
+}
+
+_READ_ONLY_TOOLS: Mapping[GoogleWorkspaceService, frozenset[str]] = {
+    'gmail': frozenset({'get_message', 'get_thread', 'list_drafts', 'list_labels', 'search_threads'}),
+    'drive': frozenset(
```
Comment (adtyavrdhn) [inline pydantic_ai_harness/google_workspace/_capability.py]: Do we really need to maintain a list of these tools?

Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: Agreed. Removed the maintained tool-name catalog. Google supplies tool discovery and read-only annotations; the capability only selects products and composes their connections.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            [
+                MCPToolset[AgentDepsT](_MCP_URLS[service], id=f'google-workspace-{service}', auth=auth).prefixed(
+                    service
+                )
+                for service in self._selected_services()
+            ]
+        )
+        if self.access == 'read':
+            toolset = toolset.filtered(lambda _ctx, tool_def: _is_read_only(tool_def))
+        if self.allowed_tools is not None:
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: The agreed batch API exposes the tools the provider makes available, including writes. Provider credentials determine access; applications can opt into read_only and compose approval_required(). Both docs now state that behavior and show approval composition. Keeping this consistent default is intentional; the capability does not add a separate approval policy.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+    services: GoogleWorkspaceService | Sequence[GoogleWorkspaceService]
+    """Workspace products to expose, such as `'gmail'` or `['gmail', 'calendar']`."""
+
+    _: KW_ONLY
+
+    access: GoogleWorkspaceAccess = 'read'
+    """`'read'` exposes only tools Google marks read-only; `'write'` exposes every tool."""
+
+    require_approval: bool = False
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: The agreed batch API exposes the tools the provider makes available, including writes. Provider credentials determine access; applications can opt into read_only and compose approval_required(). Both docs now state that behavior and show approval composition. Keeping this consistent default is intentional; the capability does not add a separate approval policy.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: tests/google_workspace/conftest.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+def fake_server() -> Iterator[str]:
+    """A stand-in Workspace server in a spawned process, yielding its URL once it accepts connections.
+
+    Spawn rather than fork: on Linux a forked child inherits the test's running event loop and
+    cannot start the server's own.
+    """
+    host = '127.0.0.1'
+    with socket.socket() as sock:
+        sock.bind((host, 0))
+        port: int = sock.getsockname()[1]
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply tests/google_workspace/conftest.py]: Addressed by removing the socket/process/thread test fixture. The focused tests now exercise Agent with an in-process MCP server, so this startup/teardown failure path no longer exists.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    auth: Auth | str | None = field(default=None, repr=False)
+    """A Google OAuth bearer token, or a custom `httpx.Auth`. Defaults to `$GOOGLE_ACCESS_TOKEN`."""
+
+    read_only: bool = False
+    """Expose only the tools Google marks read-only."""
+
+    def __post_init__(self) -> None:
+        """Normalize `services` to a tuple of products that have an endpoint."""
+        self.services = (self.services,) if isinstance(self.services, str) else tuple(self.services)
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: Reproduced the duplicate-tool failure through Agent. Service normalization now deduplicates while preserving order; the focused regression test confirms a repeated service executes once.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: tests/google_workspace/conftest.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            with socket.create_connection((host, port), timeout=0.2):
+                break
+        except OSError:
+            if not process.is_alive() or time.monotonic() > deadline:  # pragma: no cover
+                raise RuntimeError('the fake Google server did not start') from None
+            time.sleep(0.05)
+    try:
+        yield f'http://{host}:{port}/mcp'
+    finally:
+        process.terminate()
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply tests/google_workspace/conftest.py]: Addressed by removing the socket/process/thread test fixture. The focused tests now exercise Agent with an in-process MCP server, so this startup/teardown failure path no longer exists.

Reviewed against 718fc057c6af9511970a3eae29a7e8a4e4ca4884.

File: pydantic_ai_harness/_mcp.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    id: str,
+) -> AbstractToolset[DepsT]:
+    """Build a hosted MCP toolset with a fixed credential, or with the credential an `auth` function returns per run.
+
+    A fixed or unset `auth` is built immediately, so every run shares one connection and one identity; `build`
+    receives `None` only when `auth` is unset, to apply its environment or OAuth fallback. A function's `None`
+    omits the tools instead, so a run without a credential cannot fall back to the deployment's token or to
+    browser OAuth, and a function returning `'oauth'` is rejected because it would open a browser on the host.
+    """
+    if not callable(auth):
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/_mcp.py]: Fixed in 3e1e8a1c: `per_run_auth` checks `isinstance(auth, Auth)` before `callable(auth)`, so an `httpx.Auth` subclass that defines `__call__` stays a fixed credential. `test_callable_auth_object_is_fixed` in `tests/test_mcp_auth.py` covers it.

File: pydantic_ai_harness/_mcp.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+def credential(auth: str | None, *, env: str | None, service: str) -> str:
+    """The API key or token to connect with: `auth`, else the `env` variable. An empty string counts as unset."""
+    if not auth and env is not None:
+        auth = environ.get(env)
+    if not auth:
+        raise UserError(
+            f'Set `{env}` or pass `auth` to connect to {service}.' if env else f'Pass `auth` to connect to {service}.'
+        )
+    return auth
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/_mcp.py]: Addressed. A per-user `auth` function that returns `'oauth'` now raises `UserError("The \`auth\` function must return an API key or token, not 'oauth'.")`. That is the multi-tenant path, where a browser login would block a server run. A fixed `auth='oauth'` stays available on purpose, for local use. It is FastMCP's documented browser login and only runs when a developer asks for it explicitly. Tests cover both cases.

File: pydantic_ai_harness/google_workspace/_capability.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+    read_only: bool = False
+    """Keep only the tools Google marks as read-only."""
+
+    include_instructions: bool = True
+    """Pass the servers' own instructions to the agent."""
+
+    def __post_init__(self) -> None:
+        """Normalize `services` to a tuple of products that have an endpoint, and derive the `id` from them."""
+        self.services = (self.services,) if isinstance(self.services, str) else tuple(dict.fromkeys(self.services))
```
[the remark this answers is not in the document]
Comment (adtyavrdhn) [reply pydantic_ai_harness/google_workspace/_capability.py]: Fixed in 79e1c8f6. The derived id now joins the products in sorted order, so `['gmail', 'calendar']` and `['calendar', 'gmail']` both get `google-workspace-calendar-gmail` and hit Pydantic AI's duplicate-id error. `services` keeps the order the caller gave. A new test covers it.
